plugins/telemetry-otel/src/internalClusterTest/java/org/opensearch/telemetry/tracing/InMemorySpanExporter.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.telemetry.tracing;
+
+import io.opentelemetry.sdk.common.CompletableResultCode;
+import io.opentelemetry.sdk.trace.data.SpanData;
+import io.opentelemetry.sdk.trace.export.SpanExporter;
+import org.opensearch.test.telemetry.tracing.MockSpanData;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.junit.Assert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+
+/**
+ * Writes the span to the list of MockSpanData, everytime a span is emitted.
+ */
+public class InMemorySpanExporter implements SpanExporter {
```

Comment:
The Otel has that for you already, please check https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/testing/src/main/java/io/opentelemetry/sdk/testing/exporter/InMemorySpanExporter.java

Reply:
Since we are providing exporter class to plugin through setting, we dont have handle of exporter directly. Also the list of spanItems generated are not stored in static variable here, That's why created custom InMemorySpanExporter.java.

Still we want to use this existing class, will need to override plugin getTelemetry method, something like

```
    static SpanExporter exporter;

    final Settings settings;
    /**
     * Creates IntegrationTestOTelTelemetryPlugin
     * @param settings cluster settings
     */
    public IntegrationTestOTelTelemetryPlugin(Settings settings) {
        super(settings);
        this.exporter = OTelSpanExporterFactory.create(settings);
        this.settings = settings;
    }

    /**
     * This method overrides getTelemetry() method in OTel plugin class, so we create only one instance of global OpenTelemetry
     * resetForTest() will set OpenTelemetry to null again.
     * @param telemetrySettings cluster settings
     */
    public Optional<Telemetry> getTelemetry(TelemetrySettings telemetrySettings) {
        GlobalOpenTelemetry.resetForTest();
        OTelResourceProvider.get(
            settings,
            exporter,
            ContextPropagators.create(W3CTraceContextPropagator.getInstance()),
            Sampler.alwaysOn()
        );
        return super.getTelemetry(telemetrySettings);
    }
```

Not sure if this is a clean way to do.

Reply:
That's unfortunate we have to cannot use `InMemorySpanExporter` as is, probably delegation is the simplest option:

```
public class InMemorySingletonSpanExporter implements SpanExporter {
    private static final InMemorySingletonSpanExporter INSTANCE = new InMemorySingletonSpanExporter(InMemorySpanExporter.create());

    public static InMemorySingletonSpanExporter create() {
        return INSTANCE;
    }
    
    private InMemorySingletonSpanExporter(InMemorySpanExporter delegate) {
        this.delegate = delegate;
    }
    
    @Override
    public CompletableResultCode export(Collection<SpanData> spans) {
        return delegate.export(spans);
    }

    @Override
    public CompletableResultCode flush() {
        return delegate.flush();
    }

    @Override
    public CompletableResultCode shutdown() {
        return delegate.shutdown();
    }
}

```

Reply:
I think the validation logic falls apart with this approach since you have to convert between `MockSpanData` and `SpanData`, instead you could make the `TracingValidator` generic over the span type:

```
public interface TracingValidator<T> {
    /**
     * Validates spanData and return boolean value on validation status.
     * @param spans spans emitted at any point of time.
     * @param requests requests can be search/index calls.
     */
    public List<T> validate(List<T> spans, int requests);
}
```

And have a separate validators for OTEL and mocks

Reply:
Want to avoid writing same validators for OTEL and mocks, as we want to have some common validations for both the cases. 
So Instead of MockSpanData, I can create interface ISpanData which validate method understands 
```
validate(List<ISpanData> spans, int requests);
```
But here will need two implementations for both the cases and will need to convert Otel SpanData to this custom Implementation for OTel case. Does this looks good?

Reply:
Yeah, it will add more work , probably the easiest way here would be to continue with MockSpanData conversion ....